platform/common/core/generic/vault/interceptor.go

```diff
@@ -344,27 +347,32 @@ func (i *Interceptor[V]) Equals(other interface{}, nss ...string) error {
 }
 
 func (i *Interceptor[V]) Done() {
-	i.Logger.Debugf("Done with [%s], closed [%v]", i.TxID, i.Closed)
-	if !i.Closed {
+	i.Logger.Debugf("Done with [%s], closed [%v]", i.TxID, i.IsClosed())
+	if !i.IsClosed() {
+		i.Mutex.Lock()
 		i.Closed = true
+		i.Mutex.Unlock()
 		if i.QE != nil {
 			i.QE.Done()
 		}
 	}
 }
 
 func (i *Interceptor[V]) Reopen(qe VersionedQueryExecutor) error {
-	i.Logger.Debugf("Reopen with [%s], closed [%v]", i.TxID, i.Closed)
-	if !i.Closed {
+	i.Logger.Debugf("Reopen with [%s], closed [%v]", i.TxID, i.IsClosed())
+	if !i.IsClosed() {
 		return errors.Errorf("already open")
 	}
 	i.QE = qe
```

Comment:
should we protect also QE? We might have race conditions on that as well. It might happen that Done is called on QE while a GetState is happening at the same time, no?